lib/stitches/auto_fill_test.py

```diff
@@ -0,0 +1,236 @@
+from shapely import geometry as shgeo
+from deepdiff import DeepDiff
+
+from lib.stitches import auto_fill
+
+
+def test_auto_fill_box():
+    # A box with side length 10 and lower-left corner at (10, 10)
+    shape = shgeo.box(10, 10, 20, 20, ccw=True)
+
+    actual = auto_fill(
+        shape=shape,
+        angle=0,
+        row_spacing=1,
+        end_row_spacing=1,
+        max_stitch_length=1,
+        running_stitch_length=1,
+        running_stitch_tolerance=1,
+        staggers=4,
+        skip_last=False,
+        starting_point=(10, 10),
+        ending_point=(10, 10),
+        underpath=False,
+    )
+
+    # The start and ending points should be the same
+    assert actual[0].isclose(actual[-1])
+
+    expected = [
+        (10.0, 10.0),
```

Comment:
Overall, having tests would be really valuable for this project.  The thing is, I'm not feeling great about hardcoding a bunch of points like this.  We tweak our algorithms constantly, and I'm concerned that we'll constantly have to adjust these tests.

I think I'd rather if we test attributes of the returned stitches, rather than requiring an exact match to a hardcoded list of coordinates for this one rectangular shape.

Things I can think of to test:

* Are the rows actually separated by the row spacing we provided?
* Are the stitches staggered properly?
* Are the rows at the angle specified?
* Are points repeated?
  * (not sure if we should actually test this since we filter duplicate points later on)
* is underpathing actually covered by later rows of stitches?
  * this one seems pretty difficult to write a test case for and probably of limited usefulness

We could test a few specific shapes and perhaps even generate shapes randomly (sort of like fuzz testing).

Reply:
I thought about something like this. Obviously you want the loosest possible assertion that meets the function's contract. However, I couldn't quite determine what the right properties were to test, and even your suggestions seem to omit the main goal of the function, which is to ensure that the shape is completely filled. To do something like, say, generate a grid of points that should be covered by the output (and ignoring the direction of the path), or to identify the rows in the output, suggests re-implementing a significant subset of the functionality being tested, which is a bit odd.

Is there any existing code that extracts these sorts of properties, which is tested in isolation already, that I could use here? Or would you be open to me creating such a location of "property extractors" that could be used in tests?

Reply:
I don't understand what you're asking for...

Testing whether the fill covers the shape completely might be easy with shapely.  Convert the stitches to a LineString (the LineString constructor will happily accept a list of stitches), buffer by half the row spacing, intersect with the fill shape, and check the area of the intersection.  If > 99% (or something) declare victory.